cEP0002.md

```diff
@@ -0,0 +1,272 @@
+# Next Gen Bear Design
+
+| Metadata     |                                         |
+| ------------ |-----------------------------------------|
+| cEP          | 2                                       |
+| Version      | 0.1                                     |
+| Title        | Next gen bear design                    |
+| Authors      | Udayan Tandon <udayan12167@iiitd.ac.in> |
+| Status       | Proposed                                |
+| Type         | Process                                 |
+
+# Abstract
+
+This document describes how the next generation bears will be designed and
+implemented.
+
+This cEP describes a new implementation of bears in coala which should make
+ them simpler and make their design more intuitive and better.
+
+# Introduction
+
+Currently the bear design is clunky and bad. The current design makes global
+bears dependant on local bears through it's queuing mechanism. It uses
+communication signals over a queue to indicate that global bears are done with
+their execution and local bears can begin. The queueing mechanism doesn't work
+very well as it sometimes has synchronization problems. Also if local bears
+fail to run it stalls the global bears. Also we have a permanent logger thread.
+An extra logger thread consumes time and resources of the system which we can
+avoid.
+
+The new proposal aims to eliminate the use of the queuing control signals.
+Utilising a multiprocessing pool to generate the results from the bears. No
+longer will the local bears be dependent on global bears to finish. The logging
+will also be handled by the main thread itself instead of utilising an extra
+thread.
+
+# Proposed Change
+
+Here is the detailed implementation stepwise:
+
+1. coala will have a core multiprocessing pool.
```

Comment:
after discussing with @Makman2 we'll probably want to go the ProcessExecutor way which is a higher abstraction